src/allmydata/storage_client.py

```diff
@@ -254,6 +257,117 @@ def get_longname(self):
     def get_nickname(self):
         return "?"
 
+class _AnonymousStorage(object):
+    """
+    Abstraction for connecting to an anonymous storage server.
+    """
+    @classmethod
+    def from_announcement(cls, server_id, ann):
+        """
+        Create an instance from an announcement like::
+
+            {"anonymous-storage-FURL": "pb://...@...",
+             "permutation-seed-base32": "...",
+             "nickname": "...",
+            }
+
+        *nickname* is optional.
+        """
+        self = cls()
+        self._furl = str(ann["anonymous-storage-FURL"])
+        m = re.match(r'pb://(\w+)@', self._furl)
+        assert m, self._furl
+        tubid_s = m.group(1).lower()
+        self._tubid = base32.a2b(tubid_s)
+        if "permutation-seed-base32" in ann:
+            ps = base32.a2b(str(ann["permutation-seed-base32"]))
+        elif re.search(r'^v0-[0-9a-zA-Z]{52}$', server_id):
+            ps = base32.a2b(server_id[3:])
+        else:
+            log.msg("unable to parse serverid '%(server_id)s as pubkey, "
+                    "hashing it to get permutation-seed, "
+                    "may not converge with other clients",
+                    server_id=server_id,
+                    facility="tahoe.storage_broker",
+                    level=log.UNUSUAL, umid="qu86tw")
+            ps = hashlib.sha256(server_id).digest()
+        self._permutation_seed = ps
+
+        assert server_id
+        self._long_description = server_id
+        if server_id.startswith("v0-"):
+            # remove v0- prefix from abbreviated name
+            self._short_description = server_id[3:3+8]
+        else:
+            self._short_description = server_id[:8]
+        self._nickname = ann.get("nickname", "")
+        return self
+
+    def get_permutation_seed(self):
+        return self._permutation_seed
+
+    def get_name(self):
+        return self._short_description
+
+    def get_longname(self):
+        return self._long_description
+
+    def get_lease_seed(self):
+        return self._tubid
+
+    def get_tubid(self):
+        return self._tubid
+
+    def get_nickname(self):
+        return self._nickname
+
+    def connect_to(self, tub, got_connection):
+        return tub.connectTo(self._furl, got_connection)
+
+
+class _NullStorage(object):
+    """
+    Abstraction for *not* communicating with a storage server of a type with
+    which we can't communicate.
+    """
+    def get_permutation_seed(self):
+        return hashlib.sha256("").digest()
+
+    def get_name(self):
+        return "<unsupported>"
+
+    def get_longname(self):
```

Comment:
Instead of all these getters, could we declare the interface as using attributes instead? (I guess "have lots of `get_` methods is very tahoe-esque, but also un-pythonic...)

Reply:
yea that's fine with me